In [1]:
import numpy as np
import pandas as pd

In [2]:
titanic = pd.read_csv('C:/Users/Playdata/Downloads/data/titanic.csv')
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
titanic2 = pd.read_csv('C:/Users/Playdata/Downloads/data/titanic.csv')
titanic2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
titanic.index

RangeIndex(start=0, stop=891, step=1)

In [5]:
titanic.isna().sum()
# 결측치 확인하는 법.

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [6]:
titanic.Embarked.value_counts()
# 탑승항 비율 확인.

S    644
C    168
Q     77
Name: Embarked, dtype: int64

In [7]:
titanic['Embarked'] = titanic['Embarked'].fillna('S')
titanic.Cabin.isnull().sum()
# 결측치 데이터 채우기 비교 '1' / 2

687

In [8]:
titanic.Embarked.value_counts()
# 가장 많은 'S' 항을 결측치에 대입.

S    646
C    168
Q     77
Name: Embarked, dtype: int64

In [9]:
titanic.Age.value_counts()

24.00    30
22.00    27
18.00    26
19.00    25
30.00    25
         ..
55.50     1
70.50     1
66.00     1
23.50     1
0.42      1
Name: Age, Length: 88, dtype: int64

In [10]:
titanic.Sex.value_counts()

male      577
female    314
Name: Sex, dtype: int64

In [11]:
titanic['Age'].mean()

29.69911764705882

In [12]:
titanic.loc[titanic['Sex'] == 'male', ['Age']].mean()
# 남자 평균 나이

Age    30.726645
dtype: float64

In [13]:
titanic.loc[titanic['Sex'] == 'female', ['Age']].mean()
# 여자 평균 나이

Age    27.915709
dtype: float64

In [14]:
titanic[titanic['Sex'] == 'male'].mean()

PassengerId    454.147314
Survived         0.188908
Pclass           2.389948
Age             30.726645
SibSp            0.429809
Parch            0.235702
Fare            25.523893
dtype: float64

In [15]:
titanic['Pclass'].value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [16]:
titanic.loc[titanic['Pclass'] == 1 , ['Age']].mean()

Age    38.233441
dtype: float64

In [17]:
titanic.loc[titanic['Pclass'] == 2 , ['Age']].mean()

Age    29.87763
dtype: float64

In [18]:
titanic.loc[titanic['Pclass'] == 3 , ['Age']].mean()

Age    25.14062
dtype: float64

In [19]:
med_m_1 = titanic[(titanic['Sex'] == 'male') & (titanic['Pclass']==1)]['Age'].mean()

In [20]:
med_m_2 = titanic[(titanic['Sex'] == 'male') & (titanic['Pclass']==2)]['Age'].mean()

In [21]:
med_m_3 = titanic[(titanic['Sex'] == 'male') & (titanic['Pclass']==3)]['Age'].mean()

In [22]:
med_f_1 = titanic[(titanic['Sex'] == 'female') & (titanic['Pclass']==1)]['Age'].mean()

In [23]:
med_f_2 = titanic[(titanic['Sex'] == 'female') & (titanic['Pclass']==2)]['Age'].mean()

In [24]:
med_f_3 = titanic[(titanic['Sex'] == 'female') & (titanic['Pclass']==3)]['Age'].mean()

In [25]:
titanic.loc[(titanic['Age'].isnull())&(titanic['Sex']=='male')&(titanic['Pclass']==1),'Age'] =med_m_1

In [26]:
titanic.loc[(titanic['Age'].isnull())&(titanic['Sex']=='male')&(titanic['Pclass']==2),'Age'] =med_m_2

In [27]:
titanic.loc[(titanic['Age'].isnull())&(titanic['Sex']=='male')&(titanic['Pclass']==3),'Age'] =med_m_3

In [28]:
titanic.loc[(titanic['Age'].isnull())&(titanic['Sex']=='female')&(titanic['Pclass']==1),'Age'] =med_f_1

In [29]:
titanic.loc[(titanic['Age'].isnull())&(titanic['Sex']=='female')&(titanic['Pclass']==2),'Age'] =med_f_2

In [30]:
titanic.loc[(titanic['Age'].isnull())&(titanic['Sex']=='female')&(titanic['Pclass']==3),'Age'] =med_f_3

In [31]:
titanic.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [32]:
titanic.Age.isna().sum()

0

In [33]:
titanic.loc[titanic['Pclass'] == 1 , ['Age']].fillna('38.2')
inplace=True

In [34]:
titanic.loc[titanic['Pclass'] == 2 , ['Age']].fillna('29.8')
inplace=True

In [35]:
titanic.loc[titanic['Pclass'] == 3 , ['Age']].fillna('25.1')
inplace=True

In [36]:
titanic.Age.isna().sum()

0

In [37]:
titanic[(titanic['Pclass'] == 1) & 
        (titanic['Sex'] == 'female') ].fillna('38.2')
inplace=True

In [38]:
titanic[(titanic['Pclass'] == 1) & 
        (titanic['Sex'] == 'female') ].isna().sum()

PassengerId     0
Survived        0
Pclass          0
Name            0
Sex             0
Age             0
SibSp           0
Parch           0
Ticket          0
Fare            0
Cabin          13
Embarked        0
dtype: int64

In [39]:
titanic[(titanic['Pclass'] == 2) & 
        (titanic['Sex'] == 'female') ].fillna('29.8')
inplace=True

In [40]:
titanic[(titanic['Pclass'] == 2) & 
        (titanic['Sex'] == 'female') ].isna().sum()

PassengerId     0
Survived        0
Pclass          0
Name            0
Sex             0
Age             0
SibSp           0
Parch           0
Ticket          0
Fare            0
Cabin          66
Embarked        0
dtype: int64

In [41]:
titanic[(titanic['Pclass'] == 3) & 
        (titanic['Sex'] == 'female') ].fillna('25.1')
inplace=True

In [42]:
titanic[(titanic['Pclass'] == 3) & 
        (titanic['Sex'] == 'female') ].isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          138
Embarked         0
dtype: int64

In [43]:
titanic.Cabin.value_counts()

G6             4
B96 B98        4
C23 C25 C27    4
E101           3
F33            3
              ..
C45            1
A10            1
C91            1
C106           1
D11            1
Name: Cabin, Length: 147, dtype: int64

In [44]:
titanic.Cabin.isna().sum()

687

In [45]:
titanic['Cabin'].describe()
# cabin 열 확인하기.

count     204
unique    147
top        G6
freq        4
Name: Cabin, dtype: object

In [46]:
titanic['Cabin'].sample(15)

60       NaN
15       NaN
47       NaN
670      NaN
7        NaN
471      NaN
137     C123
536      B38
208      NaN
589      NaN
420      NaN
147      NaN
128    F E69
734      NaN
374      NaN
Name: Cabin, dtype: object

In [47]:
titanic['Cabin'] = titanic['Cabin'].str[:1]
# 첫번째 알파벳만 남기고.

In [48]:
titanic.drop('Cabin', axis=1, inplace=True)
# cabin 열 삭제 하기.

In [49]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [50]:
titanic.isna().sum()
# 결측치 삭제 확인.

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

In [51]:
titanic2 = pd.read_csv('C:/Users/Playdata/Downloads/data/titanic.csv')
titanic2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [52]:
titanic2['Age'].mean()

29.69911764705882

In [53]:
titanic2['Age'] = titanic2['Age'].fillna('29.69911764705882')
titanic2.Age.isnull().sum()
inplace = True

In [54]:
titanic2['Cabin'] = titanic2['Cabin'].fillna('N')
titanic2.Cabin.isnull().sum()
inplace = True

In [55]:
titanic2['Embarked'] = titanic2['Embarked'].fillna('N')
titanic2.Embarked.isnull().sum()
inplace = True

In [56]:
titanic2.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [57]:
titanic2['Cabin'] = titanic2['Cabin'].str[:1]

In [58]:
titanic.corr()

PassengerId  Survived    Pclass       Age     SibSp     Parch  \
PassengerId     1.000000 -0.005007 -0.035144  0.039636 -0.057527 -0.001652   
Survived       -0.005007  1.000000 -0.338481 -0.067485 -0.035322  0.081629   
Pclass         -0.035144 -0.338481  1.000000 -0.407015  0.083081  0.018443   
Age             0.039636 -0.067485 -0.407015  1.000000 -0.251313 -0.180705   
SibSp          -0.057527 -0.035322  0.083081 -0.251313  1.000000  0.414838   
Parch          -0.001652  0.081629  0.018443 -0.180705  0.414838  1.000000   
Fare            0.012658  0.257307 -0.549500  0.118308  0.159651  0.216225   

                 Fare  
PassengerId  0.012658  
Survived     0.257307  
Pclass      -0.549500  
Age          0.118308  
SibSp        0.159651  
Parch        0.216225  
Fare         1.000000

In [59]:
titanic2.corr()

PassengerId  Survived    Pclass     SibSp     Parch      Fare
PassengerId     1.000000 -0.005007 -0.035144 -0.057527 -0.001652  0.012658
Survived       -0.005007  1.000000 -0.338481 -0.035322  0.081629  0.257307
Pclass         -0.035144 -0.338481  1.000000  0.083081  0.018443 -0.549500
SibSp          -0.057527 -0.035322  0.083081  1.000000  0.414838  0.159651
Parch          -0.001652  0.081629  0.018443  0.414838  1.000000  0.216225
Fare            0.012658  0.257307 -0.549500  0.159651  0.216225  1.000000

In [60]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [61]:
titanic2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38      1   
2                             Heikkinen, Miss. Laina  female  26      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35      1   
4                           Allen, Mr. William Henry    male  35      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500     N        S  
1      0          PC 17599  71.2833     C        C  
2      0  STON/O2. 3101282   7.9250     N        S  
3      0            113803  53.1000     C        S  
4      0            373450   8.0500     N        S

In [62]:
titanic.drop('PassengerId', axis=1, inplace=True)
# passeng 열 삭제 하기.

In [63]:
titanic.drop('Name', axis=1, inplace=True)
# name 열 삭제 하기.

In [64]:
titanic.drop('Ticket', axis=1, inplace=True)
# ticket 열 삭제 하기

In [65]:
titanic2.drop(['PassengerId','Name','Ticket'], axis=1, inplace=True)
# 한번에 삭제 하기.

In [66]:
titanic.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [67]:
titanic2.head()

Survived  Pclass     Sex Age  SibSp  Parch     Fare Cabin Embarked
0         0       3    male  22      1      0   7.2500     N        S
1         1       1  female  38      1      0  71.2833     C        C
2         1       3  female  26      0      0   7.9250     N        S
3         1       1  female  35      1      0  53.1000     C        S
4         0       3    male  35      0      0   8.0500     N        S

# 데이터 인코딩
 + 레이블 인코딩
 + 원 핫 인코딩

In [68]:
from sklearn.preprocessing import LabelEncoder

In [69]:
from sklearn.preprocessing import LabelEncoder

items=['TV', '냉장고', '전자레인지', '컴퓨터', '선풍기', '선풍기', '믹서', '믹서']

# LabelEncoder를 객체로 생성한 후, fit()과 transform()으로 레이블 인코딩 수행

encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)

print('인코딩 변환값:', labels)
# 인코딩 변환값: [0 1 4 5 3 3 2 2]

print('인코딩 클래스:' , encoder.classes_)
# 인코딩 클래스: ['TV' '냉장고' '믹서' '선풍기' '전자레인지' '컴퓨터’]

print('디코딩 원본 값:', encoder.inverse_transform([4,5,2,0,1,1,3,3]))
# 디코딩 원본 값: ['전자레인지' '컴퓨터' '믹서' 'TV' '냉장고' '냉장고' '선풍기' '선풍기']


인코딩 변환값: [0 1 4 5 3 3 2 2]
인코딩 클래스: ['TV' '냉장고' '믹서' '선풍기' '전자레인지' '컴퓨터']
디코딩 원본 값: ['전자레인지' '컴퓨터' '믹서' 'TV' '냉장고' '냉장고' '선풍기' '선풍기']


In [70]:
aaa = titanic['Embarked']
pd.get_dummies(aaa).head(11)

C  Q  S
0   0  0  1
1   1  0  0
2   0  0  1
3   0  0  1
4   0  0  1
5   0  1  0
6   0  0  1
7   0  0  1
8   0  0  1
9   1  0  0
10  0  0  1

In [71]:
titanic.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare',
       'Embarked'],
      dtype='object')

In [72]:
bbb = titanic[['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare','Embarked']]
pd.get_dummies(bbb)

Survived  Pclass    Age  SibSp  Parch     Fare  Sex_female  Sex_male  \
0           0       3  22.00      1      0   7.2500           0         1   
1           1       1  38.00      1      0  71.2833           1         0   
2           1       3  26.00      0      0   7.9250           1         0   
3           1       1  35.00      1      0  53.1000           1         0   
4           0       3  35.00      0      0   8.0500           0         1   
..        ...     ...    ...    ...    ...      ...         ...       ...   
886         0       2  27.00      0      0  13.0000           0         1   
887         1       1  19.00      0      0  30.0000           1         0   
888         0       3  21.75      1      2  23.4500           1         0   
889         1       1  26.00      0      0  30.0000           0         1   
890         0       3  32.00      0      0   7.7500           0         1   

     Embarked_C  Embarked_Q  Embarked_S  
0             0           0           1  
1             1           0           0  
2             0           0           1  
3             0           0           1  
4             0           0           1  
..          ...         ...         ...  
886           0           0           1  
887           0           0           1  
888           0           0           1  
889           1           0           0  
890           0           1           0  

[891 rows x 11 columns]

In [73]:
titanic2.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin',
       'Embarked'],
      dtype='object')

In [74]:
bbb2 = titanic2[['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin','Embarked']]
pd.get_dummies(bbb2)

Survived  Pclass  SibSp  Parch     Fare  Sex_female  Sex_male  Age_0.42  \
0           0       3      1      0   7.2500           0         1         0   
1           1       1      1      0  71.2833           1         0         0   
2           1       3      0      0   7.9250           1         0         0   
3           1       1      1      0  53.1000           1         0         0   
4           0       3      0      0   8.0500           0         1         0   
..        ...     ...    ...    ...      ...         ...       ...       ...   
886         0       2      0      0  13.0000           0         1         0   
887         1       1      0      0  30.0000           1         0         0   
888         0       3      1      2  23.4500           1         0         0   
889         1       1      0      0  30.0000           0         1         0   
890         0       3      0      0   7.7500           0         1         0   

     Age_0.67  Age_0.75  ...  Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_N  \
0           0         0  ...        0        0        0        0        1   
1           0         0  ...        0        0        0        0        0   
2           0         0  ...        0        0        0        0        1   
3           0         0  ...        0        0        0        0        0   
4           0         0  ...        0        0        0        0        1   
..        ...       ...  ...      ...      ...      ...      ...      ...   
886         0         0  ...        0        0        0        0        1   
887         0         0  ...        0        0        0        0        0   
888         0         0  ...        0        0        0        0        1   
889         0         0  ...        0        0        0        0        0   
890         0         0  ...        0        0        0        0        1   

     Cabin_T  Embarked_C  Embarked_N  Embarked_Q  Embarked_S  
0          0           0           0           0           1  
1          0           1           0           0           0  
2          0           0           0           0           1  
3          0           0           0           0           1  
4          0           0           0           0           1  
..       ...         ...         ...         ...         ...  
886        0           0           0           0           1  
887        0           0           0           0           1  
888        0           0           0           0           1  
889        0           1           0           0           0  
890        0           0           0           1           0  

[891 rows x 109 columns]

In [75]:
from sklearn.preprocessing import LabelEncoder

items=['TV', '냉장고', '전자레인지', '컴퓨터', '선풍기', '선풍기', '믹서', '믹서']

# LabelEncoder를 객체로 생성한 후, fit()과 transform()으로 레이블 인코딩 수행
encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)
print('인코딩 변환값:', labels)
#인코딩 변환값: [0 1 4 5 3 3 2 2]

print('인코딩 클래스:', encoder.classes_)
#인코딩 클래스: ['TV' '냉장고' '믹서' '선풍기' '전자레인지' '컴퓨터’]

print('디코딩 원본 값:', encoder.inverse_transform([4,5,2,0,1,1,3,3]))
#디코딩 원본 값: ['전자레인지' '컴퓨터' '믹서' 'TV' '냉장고' '냉장고' '선풍기' '선풍기']

인코딩 변환값: [0 1 4 5 3 3 2 2]
인코딩 클래스: ['TV' '냉장고' '믹서' '선풍기' '전자레인지' '컴퓨터']
디코딩 원본 값: ['전자레인지' '컴퓨터' '믹서' 'TV' '냉장고' '냉장고' '선풍기' '선풍기']


In [76]:
items=titanic['Pclass']

encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)

titanic['Pclass'] = labels
titanic

Survived  Pclass     Sex    Age  SibSp  Parch     Fare Embarked
0           0       2    male  22.00      1      0   7.2500        S
1           1       0  female  38.00      1      0  71.2833        C
2           1       2  female  26.00      0      0   7.9250        S
3           1       0  female  35.00      1      0  53.1000        S
4           0       2    male  35.00      0      0   8.0500        S
..        ...     ...     ...    ...    ...    ...      ...      ...
886         0       1    male  27.00      0      0  13.0000        S
887         1       0  female  19.00      0      0  30.0000        S
888         0       2  female  21.75      1      2  23.4500        S
889         1       0    male  26.00      0      0  30.0000        C
890         0       2    male  32.00      0      0   7.7500        Q

[891 rows x 8 columns]

In [77]:
items=titanic['Sex']

encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)

titanic['Sex'] = labels
titanic

Survived  Pclass  Sex    Age  SibSp  Parch     Fare Embarked
0           0       2    1  22.00      1      0   7.2500        S
1           1       0    0  38.00      1      0  71.2833        C
2           1       2    0  26.00      0      0   7.9250        S
3           1       0    0  35.00      1      0  53.1000        S
4           0       2    1  35.00      0      0   8.0500        S
..        ...     ...  ...    ...    ...    ...      ...      ...
886         0       1    1  27.00      0      0  13.0000        S
887         1       0    0  19.00      0      0  30.0000        S
888         0       2    0  21.75      1      2  23.4500        S
889         1       0    1  26.00      0      0  30.0000        C
890         0       2    1  32.00      0      0   7.7500        Q

[891 rows x 8 columns]

In [78]:
items=titanic['Age']

encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)

titanic['Age'] = labels
titanic

Survived  Pclass  Sex  Age  SibSp  Parch     Fare Embarked
0           0       2    1   29      1      0   7.2500        S
1           1       0    0   56      1      0  71.2833        C
2           1       2    0   35      0      0   7.9250        S
3           1       0    0   52      1      0  53.1000        S
4           0       2    1   52      0      0   8.0500        S
..        ...     ...  ...  ...    ...    ...      ...      ...
886         0       1    1   37      0      0  13.0000        S
887         1       0    0   24      0      0  30.0000        S
888         0       2    0   28      1      2  23.4500        S
889         1       0    1   35      0      0  30.0000        C
890         0       2    1   46      0      0   7.7500        Q

[891 rows x 8 columns]

In [79]:
items=titanic['SibSp']

encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)

titanic['SibSp'] = labels
titanic

Survived  Pclass  Sex  Age  SibSp  Parch     Fare Embarked
0           0       2    1   29      1      0   7.2500        S
1           1       0    0   56      1      0  71.2833        C
2           1       2    0   35      0      0   7.9250        S
3           1       0    0   52      1      0  53.1000        S
4           0       2    1   52      0      0   8.0500        S
..        ...     ...  ...  ...    ...    ...      ...      ...
886         0       1    1   37      0      0  13.0000        S
887         1       0    0   24      0      0  30.0000        S
888         0       2    0   28      1      2  23.4500        S
889         1       0    1   35      0      0  30.0000        C
890         0       2    1   46      0      0   7.7500        Q

[891 rows x 8 columns]

In [80]:
items=titanic['Parch']

encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)

titanic['Parch'] = labels
titanic

Survived  Pclass  Sex  Age  SibSp  Parch     Fare Embarked
0           0       2    1   29      1      0   7.2500        S
1           1       0    0   56      1      0  71.2833        C
2           1       2    0   35      0      0   7.9250        S
3           1       0    0   52      1      0  53.1000        S
4           0       2    1   52      0      0   8.0500        S
..        ...     ...  ...  ...    ...    ...      ...      ...
886         0       1    1   37      0      0  13.0000        S
887         1       0    0   24      0      0  30.0000        S
888         0       2    0   28      1      2  23.4500        S
889         1       0    1   35      0      0  30.0000        C
890         0       2    1   46      0      0   7.7500        Q

[891 rows x 8 columns]

In [81]:
items=titanic['Fare']

encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)

titanic['Fare'] = labels
titanic

Survived  Pclass  Sex  Age  SibSp  Parch  Fare Embarked
0           0       2    1   29      1      0    18        S
1           1       0    0   56      1      0   207        C
2           1       2    0   35      0      0    41        S
3           1       0    0   52      1      0   189        S
4           0       2    1   52      0      0    43        S
..        ...     ...  ...  ...    ...    ...   ...      ...
886         0       1    1   37      0      0    85        S
887         1       0    0   24      0      0   153        S
888         0       2    0   28      1      2   131        S
889         1       0    1   35      0      0   153        C
890         0       2    1   46      0      0    30        Q

[891 rows x 8 columns]

In [82]:
items=titanic['Embarked']

encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)

titanic['Embarked'] = labels
titanic

Survived  Pclass  Sex  Age  SibSp  Parch  Fare  Embarked
0           0       2    1   29      1      0    18         2
1           1       0    0   56      1      0   207         0
2           1       2    0   35      0      0    41         2
3           1       0    0   52      1      0   189         2
4           0       2    1   52      0      0    43         2
..        ...     ...  ...  ...    ...    ...   ...       ...
886         0       1    1   37      0      0    85         2
887         1       0    0   24      0      0   153         2
888         0       2    0   28      1      2   131         2
889         1       0    1   35      0      0   153         0
890         0       2    1   46      0      0    30         1

[891 rows x 8 columns]

In [83]:
titanic.mean()

Survived      0.383838
Pclass        1.308642
Sex           0.647587
Age          40.904602
SibSp         0.507295
Parch         0.381594
Fare        105.355780
Embarked      1.536476
dtype: float64

In [84]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [85]:
scaler.fit(titanic)
First_scaled = scaler.transform(titanic)

In [86]:
First_df_scaled = pd.DataFrame(data = First_scaled, columns=list(titanic.columns))

In [87]:
print('feature들의 평균값')
print(First_df_scaled.mean())
print('\nfeature들의 분산값')
print(First_df_scaled.var())

feature들의 평균값
Survived   -2.287732e-16
Pclass     -2.031048e-16
Sex        -4.059603e-16
Age         5.195993e-17
SibSp       4.366130e-16
Parch       6.716164e-17
Fare       -8.398320e-17
Embarked    1.556306e-16
dtype: float64

feature들의 분산값
Survived    1.001124
Pclass      1.001124
Sex         1.001124
Age         1.001124
SibSp       1.001124
Parch       1.001124
Fare        1.001124
Embarked    1.001124
dtype: float64


In [88]:
from sklearn.preprocessing import StandardScaler

# StandardScaler객체 생성
scaler = StandardScaler()

# StandardScaler로 데이터 세트 변환. fit()과 transform() 호출
scaler.fit(titanic)
First_scaled = scaler.transform(titanic)

# transform() 시 스케일 변환된 데이터 세트가 NumPy ndarray로 반환돼 이를 DataFrame로 변환
First_df_scaled = pd.DataFrame(data=First_scaled, columns=list(titanic.columns))
print('feature들의 평균값')
print(First_df_scaled.mean())
print('\nfeature들의 분산값')
print(First_df_scaled.var())


feature들의 평균값
Survived   -2.287732e-16
Pclass     -2.031048e-16
Sex        -4.059603e-16
Age         5.195993e-17
SibSp       4.366130e-16
Parch       6.716164e-17
Fare       -8.398320e-17
Embarked    1.556306e-16
dtype: float64

feature들의 분산값
Survived    1.001124
Pclass      1.001124
Sex         1.001124
Age         1.001124
SibSp       1.001124
Parch       1.001124
Fare        1.001124
Embarked    1.001124
dtype: float64


In [89]:
titanic.corr()

Survived    Pclass       Sex       Age     SibSp     Parch  \
Survived  1.000000 -0.338481 -0.543351 -0.054879 -0.026385  0.081629   
Pclass   -0.338481  1.000000  0.131900 -0.426434  0.078141  0.018443   
Sex      -0.543351  0.131900  1.000000  0.110153 -0.123164 -0.245489   
Age      -0.054879 -0.426434  0.110153  1.000000 -0.257124 -0.162008   
SibSp    -0.026385  0.078141 -0.123164 -0.257124  1.000000  0.423338   
Parch     0.081629  0.018443 -0.245489 -0.162008  0.423338  1.000000   
Fare      0.333943 -0.724119 -0.265389  0.178125  0.368688  0.361243   
Embarked -0.167675  0.162098  0.108262 -0.005586  0.065635  0.039798   

              Fare  Embarked  
Survived  0.333943 -0.167675  
Pclass   -0.724119  0.162098  
Sex      -0.265389  0.108262  
Age       0.178125 -0.005586  
SibSp     0.368688  0.065635  
Parch     0.361243  0.039798  
Fare      1.000000 -0.158036  
Embarked -0.158036  1.000000

In [90]:
titanic

Survived  Pclass  Sex  Age  SibSp  Parch  Fare  Embarked
0           0       2    1   29      1      0    18         2
1           1       0    0   56      1      0   207         0
2           1       2    0   35      0      0    41         2
3           1       0    0   52      1      0   189         2
4           0       2    1   52      0      0    43         2
..        ...     ...  ...  ...    ...    ...   ...       ...
886         0       1    1   37      0      0    85         2
887         1       0    0   24      0      0   153         2
888         0       2    0   28      1      2   131         2
889         1       0    1   35      0      0   153         0
890         0       2    1   46      0      0    30         1

[891 rows x 8 columns]

In [91]:
titanic2.corr()

Survived    Pclass     SibSp     Parch      Fare
Survived  1.000000 -0.338481 -0.035322  0.081629  0.257307
Pclass   -0.338481  1.000000  0.083081  0.018443 -0.549500
SibSp    -0.035322  0.083081  1.000000  0.414838  0.159651
Parch     0.081629  0.018443  0.414838  1.000000  0.216225
Fare      0.257307 -0.549500  0.159651  0.216225  1.000000

In [92]:
y_titanic_df = titanic['Survived']
X_titanic_df = titanic.drop('Survived', axis=1)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_titanic_df, y_titanic_df, test_size=0.2, random_state=12)

print(X_train.shape)
print(X_test.shape)

print(y_train.shape)
print(y_test.shape)

# x트레인 - 712개 값 테스트는 - 179개  학습용- 712개 테스트  y값 -179개

(712, 7)
(179, 7)
(712,)
(179,)


In [93]:
titanic.drop('Survived', axis=1, inplace=False)

Pclass  Sex  Age  SibSp  Parch  Fare  Embarked
0         2    1   29      1      0    18         2
1         0    0   56      1      0   207         0
2         2    0   35      0      0    41         2
3         0    0   52      1      0   189         2
4         2    1   52      0      0    43         2
..      ...  ...  ...    ...    ...   ...       ...
886       1    1   37      0      0    85         2
887       0    0   24      0      0   153         2
888       2    0   28      1      2   131         2
889       0    1   35      0      0   153         0
890       2    1   46      0      0    30         1

[891 rows x 7 columns]

총데이터는 891개 중에서 0.2(랜덤)을 뽑아두고. 나머지 0.8을 두고 
712개의 입력값, 같은 값을 가진게 7개
179개의 결과값.

712개의 테스트할 결과값 과 179개는 테스트할값.

In [94]:
titanic

Survived  Pclass  Sex  Age  SibSp  Parch  Fare  Embarked
0           0       2    1   29      1      0    18         2
1           1       0    0   56      1      0   207         0
2           1       2    0   35      0      0    41         2
3           1       0    0   52      1      0   189         2
4           0       2    1   52      0      0    43         2
..        ...     ...  ...  ...    ...    ...   ...       ...
886         0       1    1   37      0      0    85         2
887         1       0    0   24      0      0   153         2
888         0       2    0   28      1      2   131         2
889         1       0    1   35      0      0   153         0
890         0       2    1   46      0      0    30         1

[891 rows x 8 columns]

In [95]:
from sklearn.tree import DecisionTreeClassifier

from sklearn.ensemble import RandomForestClassifier

from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score

# 결정트리, Random forest, 로지스틱 회귀를 위한 사이킷런 Classifier 클래스 생성

dt_clf = DecisionTreeClassifier(random_state=11)

rf_clf = RandomForestClassifier(random_state=11)

lr_clf = LogisticRegression()


In [96]:
# fit 학습 predict  accuracy_score

In [97]:
dt_clf.fit(X_train,y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=11, splitter='best')

In [98]:
dt_pred=dt_clf.predict(X_test)

In [99]:
print('DecisionTreeClassifier 정확도 : {0:2f}'.format(accuracy_score(y_test, dt_pred)))

DecisionTreeClassifier 정확도 : 0.754190


In [100]:
X_train

Pclass  Sex  Age  SibSp  Parch  Fare  Embarked
8         2    0   37      0      2    74         2
150       1    1   72      0      0    82         2
221       1    1   37      0      0    85         2
365       2    1   42      0      0    18         2
324       2    1   36      6      2   205         2
..      ...  ...  ...    ...    ...   ...       ...
241       2    0   28      1      0   102         1
253       2    1   42      1      0   109         2
390       0    1   53      1      2   232         2
667       2    1   36      0      0    31         2
843       2    1   50      0      0     4         0

[712 rows x 7 columns]

In [101]:
y_test

456    0
351    0
173    0
671    0
836    0
      ..
877    0
35     0
814    0
363    0
871    1
Name: Survived, Length: 179, dtype: int64

In [102]:
from sklearn.model_selection import KFold

def exec_kfold(clf, folds=5):
    # 폴드 세트를 5개인 KFold객체를 생성, 폴드 수만큼 예측결과 저장을 위한 리스트 객체 생성.
    kfold = KFold(n_splits=folds)
    scores = []
    # KFold 교차 검증 수행.
    for iter_count, (train_index, test_index) in enumerate(kfold.split(X_titanic_df)):
        # X_titanic_df 데이터에서 교차 검증별로 학습과 검증 데이터를 가리키는 index 생성
        X_train, X_test = X_titanic_df.values[train_index], X_titanic_df.values[test_index]
        y_train, y_test = y_titanic_df.values[train_index], y_titanic_df.values[test_index]
        # Classifier 학습, 예측, 정확도 계산
        clf.fit(X_train, y_train)
        predictions = clf.predict(X_test)
        accuracy = accuracy_score(y_test, predictions)
        scores.append(accuracy)
        print("교차 검증 {0} 정확도: {1:.4f}".format(iter_count, accuracy))


In [103]:
from sklearn.model_selection import KFold

def exec_kfold(clf, folds=5):
    # 폴드 세트를 5개인 KFold객체를 생성, 폴드 수만큼 예측결과 저장을 위한 리스트 객체 생성.
    kfold = KFold(n_splits=folds)
    scores = []
    # KFold 교차 검증 수행.
    for iter_count, (train_index, test_index) in enumerate(kfold.split(X_titanic_df)):
        # X_titanic_df 데이터에서 교차 검증별로 학습과 검증 데이터를 가리키는 index 생성
        X_train, X_test = X_titanic_df.values[train_index], X_titanic_df.values[test_index]
        y_train, y_test = y_titanic_df.values[train_index], y_titanic_df.values[test_index]
        # Classifier 학습, 예측, 정확도 계산
        clf.fit(X_train, y_train)
        predictions = clf.predict(X_test)
        accuracy = accuracy_score(y_test, predictions)
        scores.append(accuracy)
        
    mean_score = np.mean(scores)
    print("평균 정확도: {0:.4f}".format(mean_score))


In [104]:
# exec_kfold 호출
exec_kfold(dt_clf, folds=5)

# exec_kfold 호출
exec_kfold(rf_clf, folds=5)

# exec_kfold 호출
exec_kfold(lr_clf, folds=5)


평균 정확도: 0.7767
평균 정확도: 0.8137
평균 정확도: 0.7980


C:\Users\Playdata\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\Playdata\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logist

만약 1의 분포가 5% 내외라서 일부 검증 레이블 데이터 분포가 극단적이라면?
ex) 검증 레이블 데이터 분포
             0 200
             1   0
    

In [105]:
from sklearn.model_selection import cross_val_score

for i in [dt_clf, rf_clf, lr_clf]:
    scores = cross_val_score(i, X_titanic_df, y_titanic_df, cv=5)
    for iter_count, accuracy in enumerate(scores):
        print("교차 검증 {0} 정확도: {1:.4f}".format(iter_count, accuracy))
    print("평균 정확도: {0:.4f}".format(np.mean(scores)))

교차 검증 0 정확도: 0.7318
교차 검증 1 정확도: 0.7865
교차 검증 2 정확도: 0.8202
교차 검증 3 정확도: 0.7697
교차 검증 4 정확도: 0.7865
평균 정확도: 0.7790
교차 검증 0 정확도: 0.7765
교차 검증 1 정확도: 0.7978
교차 검증 2 정확도: 0.8539
교차 검증 3 정확도: 0.7809
교차 검증 4 정확도: 0.8315
평균 정확도: 0.8081
교차 검증 0 정확도: 0.7821
교차 검증 1 정확도: 0.8090
교차 검증 2 정확도: 0.7865
교차 검증 3 정확도: 0.7921
교차 검증 4 정확도: 0.8146
평균 정확도: 0.7969


C:\Users\Playdata\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
C:\Users\Playdata\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logist

In [106]:
titanic

Survived  Pclass  Sex  Age  SibSp  Parch  Fare  Embarked
0           0       2    1   29      1      0    18         2
1           1       0    0   56      1      0   207         0
2           1       2    0   35      0      0    41         2
3           1       0    0   52      1      0   189         2
4           0       2    1   52      0      0    43         2
..        ...     ...  ...  ...    ...    ...   ...       ...
886         0       1    1   37      0      0    85         2
887         1       0    0   24      0      0   153         2
888         0       2    0   28      1      2   131         2
889         1       0    1   35      0      0   153         0
890         0       2    1   46      0      0    30         1

[891 rows x 8 columns]

In [109]:
from sklearn.model_selection import GridSearchCV

parameters1 = {'max_depth':[1,2,3,4,5,6,7,8,9,10],
             'min_samples_split':[2,3,4,5,6,7,8,9,10], 'min_samples_leaf':[1,2,3,4,5,6,7,8,9,10]}

grid_dclf = GridSearchCV(dt_clf, param_grid=parameters1, scoring='accuracy', cv=5)
grid_dclf.fit(X_train, y_train)

print('GridSearchCV 최적 하이퍼 파라미터 :', grid_dclf.best_params_)
print('GridSearchCV 최고 정확도: {0:.4f}'.format(grid_dclf.best_score_))

best_dclf = grid_dclf. best_estimator_

# GridSEarchCV의 최적 하이퍼 파라미터로 학습된 Estimator로 예측 및 평가 수행

dpredictions = best_dclf.predict(X_test)
accuracy = accuracy_score(y_test, dpredictions)

print('테스트 세트에서의 DecisionTreeClassifier 정확도 : {0:.4f}'.format(accuracy))

GridSearchCV 최적 하이퍼 파라미터 : {'max_depth': 6, 'min_samples_leaf': 5, 'min_samples_split': 2}
GridSearchCV 최고 정확도: 0.8399
테스트 세트에서의 DecisionTreeClassifier 정확도 : 0.7765
